### ISSUE TITLE
Disable libtbx testing for dials

### ISSUE THREAD

**Issue by Anthchirp** [dials/dials#890]
We should disable the ability to run tests via `libtbx.run_tests_parallel module=dials`.

Reasons:
1. libtbx testing is entirely superseded by pytest.
2. test results are misclassified. libtbx testing only crudely recognizes skipped tests, and does not tell this information to the user. It indicates `OK` instead. Some passing tests that are perfectly fine are shown as `WARNING`. Expected failures are also shown as `OK`/`WARNING`.
3. optional tests depending on the `--regression` flag are not run, results are again shown as `OK`
4. none of the relevant explanatory text for test skipping or failure makes it into the output.
5. libtbx testing does not show any warnings/deprecationwarnings
6. none of the above is obvious to anyone running libtbx tests and people will therefore assume that "_they ran the tests_" when really they did not.

**Comment by nksauter** [comment]
In the interest of collaborating with Berkeley, fix the behavior of
libtbx.run_tests_parallel.

Nicholas K. Sauter, Ph. D.
Senior Scientist, Molecular Biophysics & Integrated Bioimaging Division
Lawrence Berkeley National Laboratory
1 Cyclotron Rd., Bldg. 33R0345
Berkeley, CA 94720
(510) 486-5713

**Comment by graeme-winter** [comment]
@nksauter this is not trivial - the reasons behind moving to pytest are to allow more complex / thorough testing and the reasons to be correctly logged - point 2 above literally means that we cannot express this sophistication in the existing testing framework. How can we fix the behaviour then?

Previous attempts to improve the libtbx tests got reverted e.g. https://github.com/cctbx/cctbx_project/pull/44

The following script (which I use to run the tests) works well and gives a better summary:

```
#!/bin/bash
modules=/Users/graeme/svn/cctbx/modules
cd /tmp
pytest --regression -n 4 ${modules}/dials
```

e.g.

```
Results (653.75s):
     904 passed
       2 xfailed
      10 skipped
```

i.e. you can see exactly what was run, failed, skipped - things which cannot be properly expressed in the libtbx framework - however we "fix" the libtbx framework it _cannot_ express this.

**Comment by graeme-winter** [comment]
Has already proven useful - some of this was cargo culted over from the libtbx tests (which fail if in a non-empty directory) - the cd /tmp is meaningless above... thanks folks for the update.

**Comment by phyy-nx** [comment]
Billy and I added known failure and 'unstable' test categories to the run
tests parallel suite last week.  Maybe these help?

See:
https://github.com/cctbx/cctbx_project/commit/70aa0c17948b9d3ee6b85b9e279bdacf4994043c

Regardless, disabling libtbx.run_tests_parallel for dials means we have to
rebuild our buildbot scripts to run the dials tests in different ways.
Working within the existing test framework and making it better seems more
useful.

If you send specific examples of problems, including test names and
expected outcomes, Billy and I can look at making run_tests_parallel
module=dials behave as expected.

**Comment by Anthchirp** [comment]
I'm aware of the newly added categories (xfail, unstable) in libtbx testing, and recently added a commit to reflect those new categories in the output.xml file. The issue is not with having those categories, but recognising the existing test output. This has always been an issue with libtbx testing, and, as @graeme-winter alluded to, one where historically changes were not welcomed.
The truth is that we have simply moved on, and there is no realistic way of libtbx testing to keep up with the feature set of pytest. For example just in the last week, [thanks to @ndevenish](https://github.com/pytest-dev/pytest-randomly/pull/187), the latest release of [pytest-randomly](https://github.com/pytest-dev/pytest-randomly) can now support scitbx flex arrays, so we can now have random *and* reproducible tests.

Having said that, properly integrating DIALS tests into buildbot is super easy and I have prepared a pull request for it. There are no changes required in the buildbot repository, all that needs to be changed is the DIALS builder in bootstrap. More details in the upcoming PR.

If you want specifically want `libtbx.run_tests_parallel` to continue working for dials then I can prepare a similar patch for that, too. This would be a patch in `libtbx.run_tests_parallel` that intercepts modules dxtbx, dials and xia2, and then runs pytest directly. Realistically this only has drawbacks though - you'd still need to go into an empty directory to run it, and you would lose the ability to use custom command line parameters.

**Comment by asmit3** [comment]
Just to understand the new testing mechanism - it seems we now type ```libtbx.pytest``` within the dials repository ? Is there any way of specifying the number of processes to run in parallel to speed things up ? Previously we could specify nproc as an argument to libtbx.run_tests_parallel

**Comment by graeme-winter** [comment]
`libtbx.pytest -n N`

**Comment by graeme-winter** [comment]
Also you can use `auto`:

```
  -n numprocesses, --numprocesses=numprocesses
                        shortcut for '--dist=load --tx=NUM*popen', you can use
                        'auto' here for auto detection CPUs number on host
                        system
```

From `pytest --help`

**Comment by asmit3** [comment]
I was getting this when I tried both options previously in the dials repo
```
-bash-4.2$ libtbx.pytest -n 64
ERROR: usage: pytest [options] [file_or_dir] [file_or_dir] [...]
pytest: error: unrecognized arguments: -n
  inifile: /net/dials/raid1/asmit/ExaFEL/dials/modules/dials/pytest.ini
  rootdir: /net/dials/raid1/asmit/ExaFEL/dials/modules/dials
```

**Comment by graeme-winter** [comment]
```
Grey-Area build :) $ libtbx.pytest --version
This is pytest version 4.6.5, imported from /Users/graeme/svn/cctbx/base/Python.framework/Versions/2.7/lib/python2.7/site-packages/pytest.pyc
setuptools registered plugins:
  pytest-xdist-1.26.1 at /Users/graeme/svn/cctbx/base/Python.framework/Versions/2.7/lib/python2.7/site-packages/xdist/plugin.py
  pytest-xdist-1.26.1 at /Users/graeme/svn/cctbx/base/Python.framework/Versions/2.7/lib/python2.7/site-packages/xdist/looponfail.py
  dials-data-2.0.15 at /Users/graeme/svn/cctbx/base/Python.framework/Versions/2.7/lib/python2.7/site-packages/dials_data/pytest11.py
  pytest-sugar-0.9.2 at /Users/graeme/svn/cctbx/base/Python.framework/Versions/2.7/lib/python2.7/site-packages/pytest_sugar.py
  pytest-forked-0.2 at /Users/graeme/svn/cctbx/base/Python.framework/Versions/2.7/lib/python2.7/site-packages/pytest_forked/__init__.pyc
  pytest-mock-1.10.4 at /Users/graeme/svn/cctbx/base/Python.framework/Versions/2.7/lib/python2.7/site-packages/pytest_mock.py
```

and you get?

**Comment by rjgildea** [comment]
You need to install the pytest-xdist plugin: https://docs.pytest.org/en/3.0.1/xdist.html

**Comment by Anthchirp** [comment]
Should have been installed via bootstrap since [March 2018](https://github.com/cctbx/cctbx_project/commit/396973ced5e9ffc0cfbeca734c367c017cc7d00c)

**Comment by asmit3** [comment]
Yes, it is pytest-xdist that needs to be installed. When I did ```libtbx.show_python_dependancies``` , it didn't give any indication that pytest-xdist needs to be installed. 
@Anthchirp I usually update my builds using git pull or running the bootstrap file that was used the first time to make the build. This clearly did not get picked up through that mechanism.

**Comment by asmit3** [comment]
@graeme-winter for the record this is what I got
```
(myEnv) -bash-4.2$ libtbx.pytest --version
This is pytest version 4.4.0, imported from /net/dials/raid1/asmit/ExaFEL/dials/newconda/envs/myEnv/lib/python2.7/site-packages/pytest.pyc
setuptools registered plugins:
  pytest-mock-1.10.4 at /net/dials/raid1/asmit/ExaFEL/dials/newconda/envs/myEnv/lib/python2.7/site-packages/pytest_mock.py
  dials-data-2.0.15 at /net/dials/raid1/asmit/ExaFEL/dials/newconda/envs/myEnv/lib/python2.7/site-packages/dials_data/pytest11.py
```

**Comment by Anthchirp** [comment]
We don't list `pytest-xdist` as a dependency, as -strictly- it is not required to run the software or the tests. But still useful for developers to have. The other (in my opinion) useful pytest plugin is `pytest-sugar`. Others may come up in the future (`pytest-randomly` comes to mind).

I guess it would be useful to split up the dependencies as _"These are the things you need to run dials"_ and _"These are the things you need to develop dials"_

**Comment by rjgildea** [comment]
Whilst were on the subject of useful pytest plugins, [pytest-randomly](https://pypi.org/project/pytest-random-order/) is useful to help ensure that there are order-dependencies in tests.
